File: codes/javascript/src/operator_optional_chaining.js
console.log('\n// Optional Chaining Operator (?.) in JavaScript');

const JSON_OBJECT = {
    "foo": {
        "bar": "baz",
    }
};
console.log(`JSON_OBJECT: ${JSON.stringify(JSON_OBJECT, null, ' '.repeat(4))}`);

console.log('JSON_OBJECT?.foo?.bar:', JSON_OBJECT?.foo?.bar);
// JSON_OBJECT?.foo?.bar: baz

console.log('JSON_OBJECT?.foo?.baz:', JSON_OBJECT?.foo?.baz);
// JSON_OBJECT?.foo?.baz: undefined

File: codes/ruby/operator_optional_chaining.rb
require 'json'

print("\n# JavaScript-like Optional Chaining Operator (?.) in Ruby")

JSON_OBJECT = {
    "foo" => {
        "bar" => "baz"
    }
}
print("\nJSON_OBJECT: ", JSON.pretty_generate(JSON_OBJECT, { "indent": " " * 4 }))

print("\n# using Ruby Optional Chaining Operator (?.) built-in function \"dig\"")

print("\nJSON_OBJECT?.foo?.bar: ", JSON_OBJECT.dig("foo", "bar"))
# JSON_OBJECT?.foo?.bar: baz

print("\nJSON_OBJECT?.foo?.baz: ", JSON_OBJECT.dig("foo", "baz"))
# JSON_OBJECT?.foo?.baz: nil
